server/src/main/java/io/crate/expression/tablefunctions/MatchesFunction.java

```diff
@@ -0,0 +1,236 @@
+import io.crate.types.DataTypes;
+import io.crate.types.RowType;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.regex.Matcher;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+public class MatchesFunction extends TableFunctionImplementation<List<Object>> {
+
+    public static final String NAME = "regexp_matches";
+    private static final RowType ROW_TYPE = new RowType(
+        List.of(DataTypes.STRING_ARRAY), List.of("groups"));
+    private static final Iterator<Row> NO_RESULTS_ITERATOR = Collections.emptyIterator();
+
+    public static void register(TableFunctionModule module) {
+        module.register(
+            Signature.table(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING_ARRAY.getTypeSignature()
+            ),
+            MatchesFunction::new
+        );
+        module.register(
+            Signature.table(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING_ARRAY.getTypeSignature()
+            ),
+            MatchesFunction::new
+        );
+    }
+
+    @Nullable
+    private final RegexMatcher regexMatcher;
+    private final Signature signature;
+    private final Signature boundSignature;
+
+    private MatchesFunction(Signature signature, Signature boundSignature) {
+        this(signature, boundSignature, null);
+    }
+
+    private MatchesFunction(Signature signature, Signature boundSignature, @Nullable RegexMatcher regexMatcher) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+        this.regexMatcher = regexMatcher;
+    }
+
+    @Override
+    public Signature signature() {
+        return signature;
+    }
+
+    @Override
+    public Signature boundSignature() {
+        return boundSignature;
+    }
+
+    @Override
+    public RowType returnType() {
+        return ROW_TYPE;
+    }
+
+    @Override
+    public boolean hasLazyResultSet() {
+        return false;
+    }
+
+    @VisibleForTesting
+    RegexMatcher regexMatcher() {
+        return regexMatcher;
+    }
+
+    @Override
+    public Scalar<Iterable<Row>, List<Object>> compile(List<Symbol> arguments) {
+        assert arguments.size() > 1 : "number of arguments must be > 1";
+        String pattern = null;
+        if (arguments.get(1).symbolType() == SymbolType.LITERAL) {
+            Literal<String> literal = (Literal<String>) arguments.get(1);
+            String patternVal = literal.value();
```

Comment:
is this variable really necessary `patternVal`, you can probably just use `pattern`.